### PR TITLE
[MXNET-349] Histogram Operator

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
This PR implements histogram operator.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Implementation of internal _histogram operator
- [x] User-facing Python interface
- [x] Unit tests

## Comments ##
The user-facing interface resembles basic behaviors of Numpy's histogram: https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.histogram.html.
Three versions are supported:
histogram(x, bins=7) - divided into 7 equal bins in the range of [x.min, x.max]
histogram(x, bins=7, range=(a,b)) - divided into 7 equal bins in the range of [a, b], anything outside the range is ignored
histogram(x, bins=NDArray) - bin bounds are specified by the bins NDArray.

### CODE REVIEW

File: python/mxnet/ndarray/ndarray.py [hunk 1]
```diff
@@ -3740,3 +3740,29 @@ def empty(shape, ctx=None, dtype=None):
...
+        Values outside the range are ignored. The first element of the range must be less than or
+        equal to the second. range affects the automatic bin computation as well, the range will
+        be equally divided by the number of bins.
+    """
+
+    if isinstance(bins, NDArray):
+        return _internal._histogram(data=a, bins=bins)
+    elif isinstance(bins, int):
+        if range_ is None:
+            range_ = (float(a.min().asnumpy()[0]), float(a.max().asnumpy()[0]))
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: don't do this in the frontend.

File: python/mxnet/ndarray/ndarray.py [hunk 2]
```diff
@@ -3740,3 +3740,29 @@ def empty(shape, ctx=None, dtype=None):
...
+        return _internal._histogram(data=a, bins=array([]), bin_cnt=bins, range=range_)
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: why?

Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: Which part you're asking about?

File: src/operator/tensor/histogram.cc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{"_backward_histogram"})
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::pair<int, int> >{};
+  })
+.add_argument("data", "NDArray-or-Symbol", "Input ndarray")
+.add_argument("bins", "NDArray-or-Symbol", "Input ndarray")
+.add_arguments(HistogramParam::__FIELDS__());
+
+NNVM_REGISTER_OP(_backward_histogram)
```
Comment (piiswrong) [inline src/operator/tensor/histogram.cc]: Historgram op is not differentiable. Don't register backward op.

Comment (haojin2) [reply src/operator/tensor/histogram.cc]: Sure, will get rid of it.

File: python/mxnet/ndarray/ndarray.py [hunk 4]
```diff
@@ -3740,3 +3740,32 @@ def empty(shape, ctx=None, dtype=None):
     if dtype is None:
         dtype = mx_real_t
     return NDArray(handle=_new_alloc_handle(shape, ctx, False, dtype))
+
+
+def histogram(a, bins=10, range_=None):
```
Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: Why underscore suffix for `range`?

File: python/mxnet/ndarray/ndarray.py [hunk 5]
```diff
@@ -3740,3 +3740,32 @@ def empty(shape, ctx=None, dtype=None):
...
+        including the rightmost edge, allowing for non-uniform bin widths.
+    range_ : (float, float), optional
+        The lower and upper range of the bins. If not provided, range is simply (a.min(), a.max()).
+        Values outside the range are ignored. The first element of the range must be less than or
+        equal to the second. range affects the automatic bin computation as well, the range will
+        be equally divided by the number of bins.
+    """
+
+    # pylint: disable= no-member, protected-access
+    if isinstance(bins, NDArray):
```
Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: Need to check whether `bins` is a 1d array?

File: python/mxnet/ndarray/ndarray.py [hunk 6]
```diff
@@ -3740,3 +3740,32 @@ def empty(shape, ctx=None, dtype=None):
...
+        return _internal._histogram(data=a, bins=bins)
+    elif isinstance(bins, int):
```
Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: Replace `int` with `integer_types`.

File: tests/python/unittest/test_operator.py [hunk 7]
```diff
@@ -5919,6 +5919,31 @@ def f(x, a, b, c):
         check_numeric_gradient(quad_sym, [data_np], atol=0.001)
 
 
+@with_seed()
+def test_histogram():
+    def f(x, bins=10, range=None):
+        return np.histogram(x, bins, range=range)
+
+    shape = (3, 3, 3)
```
Comment (reminisce) [inline tests/python/unittest/test_operator.py]: Randomize shape and test ndarrays from 1D-5D.

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -5919,6 +5919,31 @@ def f(x, a, b, c):
...
+    x = rand_ndarray(shape, stype='default')
+    mx_bins = mx.nd.array([-1.0, 0.5, 2.0, 4.5, 50.0])
+    np_bins = mx_bins.asnumpy()
+    bin_cnt = 5
```
Comment (reminisce) [inline tests/python/unittest/test_operator.py]: Should test more settings of `bin_cnt`.

File: python/mxnet/ndarray/ndarray.py (hunk 2 above)
Comment (reminisce) [reply python/mxnet/ndarray/ndarray.py]: You can define a `mode` variable in `HistogramParam` for distinguishing the cases of 1 and 2 ndarray inputs. When register the op, use `set_num_inputs([](const NodeAttrs& attrs)` to define the number of inputs based upon the `mode`.
See example:
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/convolution.cc#L471

File: src/operator/tensor/histogram-inl.h [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    TShape oshape = (*in_attrs)[1];
+
+    CHECK_EQ(oshape.ndim(), 1U) << "bins argument should be an 1D vector";
+    CHECK_GE(oshape.Size(), 2U) << "number of bounds should be >= 2";
+
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape({(oshape[0] - 1)}));
+    SHAPE_ASSIGN_CHECK(*out_attrs, 1, in_attrs->at(1));
+  }
+
+  return out_attrs->at(0).ndim() == 1U && out_attrs->at(0).Size() != 0U &&
```
Comment (reminisce) [inline src/operator/tensor/histogram-inl.h]: Call `shape_is_none` to simplify the code. `ndim == 1` should have been guaranteed before this line.

File: src/operator/tensor/histogram-inl.h [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+inline bool HistogramOpType(const nnvm::NodeAttrs& attrs,
+                            std::vector<int>* in_attrs,
+                            std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(in_attrs->at(0), in_attrs->at(1));
+  CHECK_EQ(out_attrs->size(), 2U);
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kInt64);
+  TYPE_ASSIGN_CHECK(*out_attrs, 1, in_attrs->at(0));
+  return out_attrs->at(0) != -1 && out_attrs->at(1) != -1;
```
Comment (reminisce) [inline src/operator/tensor/histogram-inl.h]: Call `!type_is_none(out_attrs->at(i))`

File: python/mxnet/ndarray/ndarray.py (hunk 5 above)
Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: That's checked inside the op

File: python/mxnet/ndarray/ndarray.py (hunk 4 above)
Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: pylint complains if we override the keyword ``range``

Comment (reminisce) [reply python/mxnet/ndarray/ndarray.py]: @piiswrong `numpy.histogram` has parameter `range`. Should we disable pylint check here to make `mx.nd.histogram` parameter naming consistent with numpy?

File: src/operator/tensor/histogram-inl.h (hunk 9 above)
Comment (haojin2) [reply src/operator/tensor/histogram-inl.h]: Done

File: src/operator/tensor/histogram-inl.h (hunk 10 above)
Comment (haojin2) [reply src/operator/tensor/histogram-inl.h]: Done

File: tests/python/unittest/test_operator.py (hunk 7 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: Done

File: tests/python/unittest/test_operator.py (hunk 8 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: Done

File: python/mxnet/ndarray/ndarray.py (hunk 2 above)
Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: Done

**Comment by haojin2** [comment]
@reminisce @piiswrong @anirudh2290 @rahul003 Unit tests added, this is ready, please give a review if you have time. Thanks!

File: python/mxnet/ndarray/ndarray.py [hunk 11]
```diff
@@ -3740,3 +3740,36 @@ def empty(shape, ctx=None, dtype=None):
...
+    if isinstance(bins, NDArray):
+        return _internal._histogram(data=a, bins=bins)
+    elif isinstance(bins, integer_types):
+        if range is None:
+            warnings.warn("range_ is not specified, using numpy's result "
+                          "to ensure consistency with numpy")
+            res, bin_bounds = np.histogram(a.asnumpy(), bins=bins)
+            return array(res), array(bin_bounds)
+        return _internal._histogram(data=a, bin_cnt=bins, range=range)
+    return None
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: why return None? Shouldn't this cause an error?

Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: Do you mean that I should throw an exception?

Comment (haojin2) [reply python/mxnet/ndarray/ndarray.py]: Added an exception for the illegal input case.

File: src/operator/tensor/histogram.cc [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    return std::vector<std::string>{"data", "bins"};
+  })
+.set_attr<FResourceRequest>("FResourceRequest",
+  [](const NodeAttrs& attrs) {
+    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
+})
+.set_attr<nnvm::FInferShape>("FInferShape", HistogramOpShape)
+.set_attr<nnvm::FInferType>("FInferType", HistogramOpType)
+.set_attr<FCompute>("FCompute<cpu>", HistogramOpForward<cpu>)
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
```
Comment (reminisce) [inline src/operator/tensor/histogram.cc]: No need to register this if it's empty.

File: src/operator/tensor/histogram.cc [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+void ComputeHistogram(const int* bin_indices, CType* out_data, size_t input_size) {
+  for (size_t i = 0; i < input_size; ++i) {
+    int target = bin_indices[i];
+    if (target >= 0) {
+      out_data[target] += 1;
+    }
+  }
+}
+
+template<typename cpu>
```
Comment (reminisce) [inline src/operator/tensor/histogram.cc]: Should this be a template specialization? Same for other places.
```cpp
template<>
void HistogramForwardImpl<cpu>()
```

File: src/operator/tensor/histogram.cc (hunk 12 above)
Comment (haojin2) [reply src/operator/tensor/histogram.cc]: Done

File: src/operator/tensor/histogram.cc (hunk 13 above)
Comment (haojin2) [reply src/operator/tensor/histogram.cc]: Done

File: src/operator/tensor/histogram.cc [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+  for (size_t i = 0; i < input_size; ++i) {
+    int target = bin_indices[i];
+    if (target >= 0) {
+      out_data[target] += 1;
+    }
+  }
+}
+
+template<>
+void HistogramForwardImpl<cpu>(const OpContext& ctx,
```
Comment (reminisce) [inline src/operator/tensor/histogram.cc]: Why change this back to using `OpContext`? If using Stream for dispatching, you can get rid of the templates.

Comment (haojin2) [reply src/operator/tensor/histogram.cc]: I'm not using OpContext, I'm just getting rid of the stream argument here.

**Comment by haojin2** [comment]
@piiswrong Should be ready for merge, please take a look when you have time, thanks!

**Comment by haojin2** [comment]
@piiswrong There's no backward pass for this function, I wonder if symbolic is necessary in this case?

**Comment by haojin2** [comment]
@piiswrong symbol is added for histogram.

**Pull request merged**